proposals/306-ipv6-happy-eyeballs.txt

```diff
@@ -0,0 +1,300 @@
+Filename: 306-ipv6-happy-eyeballs.txt
+Title: A Tor Implementation of IPv6 Happy Eyeballs
+Author: Neel Chauhan
+Created: 25-Jun-2019
+Supercedes: 299
+Status: Open
+Ticket: https://trac.torproject.org/projects/tor/ticket/29801
+
+1. Introduction
+
+   As IPv4 address space becomes scarce, ISPs and organizations will deploy
+   IPv6 in their networks. Right now, Tor clients connect to guards using
+   IPv4 connectivity by default.
+
+   When networks first transition to IPv6, both IPv4 and IPv6 will be enabled
+   on most networks in a so-called "dual-stack" configuration. This is to not
+   break existing IPv4-only applications while enabling IPv6 connectivity.
+   However, IPv6 connectivity may be unreliable and clients should be able
+   to connect to the guard using the most reliable technology, whether IPv4
+   or IPv6.
+
+   In ticket #27490, we introduced the option ClientAutoIPv6ORPort which
+   lets a client randomly choose between IPv4 or IPv6. However, this
+   random decision does not take into account unreliable connectivity
+   or falling back to the competing IP version should one be unreliable
+   or unavailable.
+
+   One way to select between IPv4 and IPv6 on a dual-stack network is a
+   so-called "Happy Eyeballs" algorithm as per RFC 8305. In one, a client
+   attempts the preferred IP family, whether IPv4 or IPv6. Should it work,
+   the client sticks with the preferred IP family. Otherwise, the client
+   attempts the alternate version. This means if a dual-stack client has
+   both IPv4 and IPv6, and IPv6 is unreliable, preferred or not, the
+   client uses IPv4, and vice versa. However, if IPv4 and IPv6 are both
+   equally reliable, and IPv6 is preferred, we use IPv6.
+
+   In Proposal 299, we have attempted a IP fallback mechanism using failure
+   counters and preferring IPv4 and IPv6 based on the state of the counters.
+   However, Prop299 was not standard Happy Eyeballs and an alternative,
+   standards-compliant proposal was requested in [P299-TRAC] to avoid issues
+   from complexity caused by randomness.
+
+   This proposal describes a Tor implementation of Happy Eyeballs and is
+   intended as a successor to Proposal 299.
+
+2. Address/Relay Selection
+
+   This section describes the necessary changes for address selection to
+   implement Prop306.
+
+2.1. Extend Info Structure Changes
+
+   To be able to handle Happy Eyeballs in Tor, we will need to modify the
+   data structures used for connections to guards, namely the extend info
+   structure.
```

Comment:
We need to list the different ways Tor clients can connect into the network:
```suggestion
   structure.

   Entry nodes are usually guards, but some clients don't use guards:
     * bootstrapping clients can connect to fallback directory mirrors or
        authorities,
     * v3 single onion services can use IPv4 or IPv6 addresses to connect
        to rendezvous points, and
     * clients can be configured to disable entry guards.
   Bridges are out of scope for this proposal, because Tor does not support
   multiple IP addresses in a single bridge line.
```